app/futurevoter/tasks.py

```
@@ -0,0 +1,94 @@
+import logging
+from datetime import date, datetime, timedelta, timezone
+
+import sentry_sdk
+from celery import shared_task
+from django.conf import settings
+
+from common.i90 import shorten_url
+from smsbot.models import Number
+
+logger = logging.getLogger("futurevoter")
+
+under_18_msg = "Happy Birthday from VoteAmerica! Remember, once you turn 18 it's time to register to vote. Have friends who are turning 18? Remind them to register to vote at https://my.voteamerica.com/fvregister"
+
+
+def is_leap_year(year):
+    return (date(year, 3, 1) - timedelta(days=1)).day == 29
+
+
+# We extract this helper so we can mock it in tests
+def send_sms(number, message):
+    number.send_sms(message)
+
+
+@shared_task
+def birthday_under_18() -> None:
```

I would combine the under/over birthday tasks, make one db call for all voters, then just do a switch on whether they should receive the general birthday sms or the 18th bday sms. While I'm not super concerned with the query performance given that this is a background task, it reduces the code maintainability significantly (1 task to maintain instead of 2) since you're doing the same thing for each voter and just changing the message (right? or did i miss something?)

---

The big difference is how we select the voters. For under-18 birthday texts, we select voters whose birthday is today. For 18th birthday texts, we select voters who are over 18 and haven't gotten the text yet -- the difference is in error handling. If the daily birthday job fails, we don't want to send a late birthday text to under-18s, but we do want to keep trying to text folks when they turn 18 if the 18th-birthday job fails.

---

> For 18th birthday texts, we select voters who are over 18 and haven't gotten the text yet
Isn't that the same thing as checking each day to see if someone turned 18? I'm not sure i understand the difference.

---

It's different if the job fails for whatever reason (db downtime, server crash, etc).

For the under-18 birthday texts, if the job fails partway through, we just let it fail. It's not a big deal if we don't send out some under-18 birthday texts, and it'd look bad if we sent those a day late.

For the 18th birthday text, if the job fails partway through, we won't get all the birthday texts out that day. But the next day, because we're selecting all users over 18 who haven't yet received the text *even if it's not still their birthday*, we'll pick up those users that we failed to send a text to yesterday, and send them the text (albeit a day late) -- I think that's desirable, because we really don't want to miss sending the "you're 18, time to register" text, even if it's late.

---

Given those differing needs, makes sense to keep them separate then!